### ISSUE TITLE
Provision trident operator within k3s 1.17.3+k3s1

### ISSUE THREAD

**Issue by jbSyTech** [NetApp/trident#399]
Just trying to get the trident operator get running on your k3s Cluster, I think it's a problem how k3s tell it's version:
# kubectl version --short
Client Version: v1.17.4+k3s1
Server Version: v1.17.4+k3s1

When I try to create the TridentProvisioner CR it is allways in Failed state

# kubectl describe tprov trident -n trident
Name:         trident
Namespace:    trident
Labels:       <none>
Annotations:  <none>
API Version:  trident.netapp.io/v1
Kind:         TridentProvisioner
Metadata:
  Creation Timestamp:  2020-05-11T15:23:50Z
  Generation:          1
  Resource Version:    1379
  Self Link:           /apis/trident.netapp.io/v1/namespaces/trident/tridentprovisioners/trident
  UID:                 8efa51a9-cee2-4111-8bb3-2b697f21654e
Spec:
  Debug:  true
Status:
  Message:  Failed to install Trident; err: could not create the Trident service account; could not create service account; yamlDocument ---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: trident-csi
  labels:
    trident_version: v20.04
    app: controller.csi.trident.netapp.io
    k8s_version: v1.17.4+k3s1
  ownerReferences:
  - apiVersion: trident.netapp.io/v1
    controller: true
    kind: TridentProvisioner
    name: trident
    uid: 8efa51a9-cee2-4111-8bb3-2b697f21654e
 was not created after 30.00 seconds
  Status:   Failed
  Version:
Events:
  Type     Reason      Age   From                        Message
  ----     ------      ----  ----                        -------
  Normal   Installing  29s   trident-operator.netapp.io  Installing Trident
  Warning  Failed      1s    trident-operator.netapp.io  Failed to install Trident; err: could not create the Trident service account; could not create service account; yamlDocument ---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: trident-csi
  labels:
    trident_version: v20.04
    app: controller.csi.trident.netapp.io
    k8s_version: v1.17.4+k3s1
  ownerReferences:
  - apiVersion: trident.netapp.io/v1
    controller: true
    kind: TridentProvisioner
    name: trident
    uid: 8efa51a9-cee2-4111-8bb3-2b697f21654e
 was not created after 30.00 seconds

When I try to create the service account with metadata.labels.k8_vsersion=v1.17.4 by hand servie account creation is working, but with metadata.labels.k8_vsersion=v1.17.4+k3s1 there is an error. Is it possible to overwrite the k8s_version lable in the TridentProvisioner CR?

**Comment by gnarl** [comment]
"The ServiceAccount “trident-csi” is invalid: metadata.labels: Invalid value: “v1.17.4+k3s1": a valid label must be an empty string or consist of alphanumeric characters, ‘-’, ‘_’ or ‘.’, and must start and end with an alphanumeric character (e.g. ‘MyValue’,  or ‘my_value’,  or ‘12345’, regex used for validation is ‘(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?’)"

It appears that K3S is using a non-standard version format. We can implement a change to detect and truncate the '+' and everything after it but it won't be available until the Trident 20.07 release.

**Comment by paepke** [comment]
Same here - running ocp which reports its version as "v1.17.1+f63db30". It happens after upgrading the ocs to a newer version - trident is still running, but the operator tries to update the  deployment with the new label with the "+" in it.
Pretty sure I cannot freshly install trident on it.

**Comment by clintonk** [comment]
@jbSyTech @paepke While we fix this issue in the operator, you might try ```tridentctl install``` instead.

**Comment by rwazar** [comment]
Can anyone confirm that this issue will be fixed in 20.07?
Thanks.

**Comment by gnarl** [comment]
@rwazar, [we are planning](https://github.com/NetApp/trident/issues/399#issuecomment-626811378) to have this fix available for the Trident 20.07 release.

**Comment by gorantornqvist** [comment]
Hi, after upgrade of openshift to 4.4.10 and all my nodes were rebooted, I appeared to hit this bug, the trident-operator log say:
```
Object creation failed." err="ServiceAccount \"trident-csi\" is invalid: metadata.labels: Invalid value: \"v1.17.1+9d33dd3\": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character ...
```
So I tried the workaround described to run ./tridentctl install -n trident but I get:
FATA Install failed; CSI Trident is already installed in namespace trident.  Resolve the issue; use 'tridentctl uninstall' to clean up; and try again.
Is the correct workaround to do "tridentctl uninstall" first?

(trident-installer 20.04.0)

**Comment by Numblesix** [comment]
> Hi, after upgrade of openshift to 4.4.10 and all my nodes were rebooted, I appeared to hit this bug, the trident-operator log say:
> 
> ```
> Object creation failed." err="ServiceAccount \"trident-csi\" is invalid: metadata.labels: Invalid value: \"v1.17.1+9d33dd3\": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character ...
> ```
> 
> So I tried the workaround described to run ./tridentctl install -n trident but I get:
> FATA Install failed; CSI Trident is already installed in namespace trident. Resolve the issue; use 'tridentctl uninstall' to clean up; and try again.
> Is the correct workaround to do "tridentctl uninstall" first?
> 
> trident-installer 20.04.0.
> 
> EDIT:
> 
> ```
> $ oc get pods -n trident
> NAME                               READY   STATUS              RESTARTS   AGE
> trident-csi-59xh8                  0/2     ContainerCreating   0          21d
> trident-csi-c2gn4                  0/2     ContainerCreating   0          21d
> trident-csi-jmzv4                  0/2     ContainerCreating   0          21d
> trident-csi-lzd24                  0/2     ContainerCreating   0          21d
> trident-csi-ppfsq                  0/2     ContainerCreating   0          21d
> trident-csi-pq2hp                  0/2     ContainerCreating   0          21d
> trident-csi-q9wc6                  0/2     ContainerCreating   0          21d
> trident-csi-qm9bt                  0/2     ContainerCreating   0          21d
> trident-csi-wnj6j                  0/2     ContainerCreating   0          21d
> trident-operator-564d7d66f-8x6fn   1/1     Running             0          119m
> 
> $ oc describe tprov trident -n trident
> Name:         trident
> Namespace:    trident
> Labels:       <none>
> Annotations:  <none>
> API Version:  trident.netapp.io/v1
> Kind:         TridentProvisioner
> Metadata:
>   Creation Timestamp:  2020-05-06T07:09:55Z
>   Generation:          1
>   Resource Version:    259179120
>   Self Link:           /apis/trident.netapp.io/v1/namespaces/trident/tridentprovisioners/trident
>   UID:                 77803ce7-f224-493c-8251-a572974104f2
> Spec:
>   Debug:  true
> Status:
>   Message:  Failed to install Trident; err: could not create the Trident service account; could not create service account; yamlDocument ---
> apiVersion: v1
> kind: ServiceAccount
> metadata:
>   name: trident-csi
>   labels:
>     app: controller.csi.trident.netapp.io
>     k8s_version: v1.17.1+9d33dd3
>     trident_version: v20.04
>   ownerReferences:
>   - apiVersion: trident.netapp.io/v1
>     controller: true
>     kind: TridentProvisioner
>     name: trident
>     uid: 77803ce7-f224-493c-8251-a572974104f2
>  was not created after 30.00 seconds
>   Status:   Failed
>   Version:  v20.04
> Events:
>   Type     Reason  Age                    From                        Message
>   ----     ------  ----                   ----                        -------
>   Warning  Failed  5h55m (x117 over 34h)  trident-operator.netapp.io  Failed to install Trident; err: could not create the Trident service account; could not create service account; yamlDocument ---
> apiVersion: v1
> kind: ServiceAccount
> metadata:
>   name: trident-csi
>   labels:
>     app: controller.csi.trident.netapp.io
>     k8s_version: v1.17.1+912792b
>     trident_version: v20.04
>   ownerReferences:
>   - uid: 77803ce7-f224-493c-8251-a572974104f2
>     apiVersion: trident.netapp.io/v1
>     controller: true
>     kind: TridentProvisioner
>     name: trident
>  was not created after 30.00 seconds
> ```

I guess one way would be to remove the old operator and redo an install using the cli .

In theory you should be able to import the old pv so you shouldnt loose any data

**Comment by gorantornqvist** [comment]
thanks ...
I scaled down the operator deployment to 0 and then did tridentctl uninstall/install ...
all pvc:s remained so no import was needed ...

**Comment by cprivite** [comment]
This is an issue with VMWare TKG installs of kubernetes as well, as their version string is of the sort of "1.18.3+vmware.1".

**Comment by gnarl** [comment]
@cprivite you should also be able to use the tridentctl install until the Trident 20.07 release is available. The "+" symbol is a character that was not supported in K8S but a number of projects started using it recently in their version strings.

**Comment by kenankule** [comment]
@gnarl I've tried with a build from master, and Pod creation (transient-trident-version-pod) fails probably due to this line: https://github.com/NetApp/trident/blob/master/operator/controllers/provisioner/installer/installer.go#L1710
Do you need a new ticket to fix that?

**Comment by gnarl** [comment]
@kenankule, thanks for the information! A new ticket isn't necessary and we'll look into the issue this week.

**Comment by gnarl** [comment]
@kenankule https://github.com/NetApp/trident/commit/24738b58e03d50b4dc9b278e124d4bbbbdae3254

**Comment by gnarl** [comment]
This issue is fixed in the Trident 20.07 release.
